### PR TITLE
Integrating Azure Data Lake Filesystem

### PR DESCRIPTION
**Pull request by jinhyukchang**
- Added:
  Runtime dependency of Azure Data Lake Filesystem
- Changed:
  Pass through job property, with ability to decrypt, to Configuration to instantiate FileSystem.
  

Pull file snippet for Dist cp case:
from=/from/directory
to=/to/directory

writer.fs.uri=adl://mydatalakestore.azuredatalakestore.net/

dfs.adls.oauth2.access.token.provider.type=ClientCredential
dfs.adls.oauth2.refresh.url=https://[end point token retrieval]
dfs.adls.oauth2.client.id=[client ID]
writer.encrypted.dfs.adls.oauth2.credential=[encrypted ADL secret key]
encrypt.key.loc=/passphrase/file/path

### CODE REVIEW

**Comment by jinhyukchang** [comment]
@ibuenros Can you take a look? Thanks.

File: gobblin-core/build.gradle [hunk 1]
```diff
@@ -57,6 +57,7 @@ dependencies {
   compile externalDependency.hadoopHdfs
 
   runtime externalDependency.protobuf
+  runtime externalDependency.hadoopAdl
```
Comment (ibuenros) [inline gobblin-core/build.gradle]: Why is this needed?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 2]
```diff
@@ -56,7 +56,7 @@
 import gobblin.publisher.DataPublisher;
 import gobblin.publisher.UnpublishedHandling;
 import gobblin.util.HadoopUtils;
-
+import gobblin.util.WriterUtils;
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Add empty line after this one

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java [hunk 3]
```diff
@@ -68,7 +69,6 @@
 import gobblin.util.limiter.Limiter;
 import gobblin.util.limiter.broker.SharedLimiterFactory;
 import gobblin.writer.DataWriter;
-
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java]: This line should not be removed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java [hunk 4]
```diff
@@ -38,6 +38,7 @@
 import com.google.common.base.Predicate;
 import com.google.common.base.Strings;
 import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java]: Is this needed?

File: gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java [hunk 5]
```diff
@@ -462,4 +464,34 @@ public static boolean verifySubset(Config superConfig, Config subConfig) {
...
+   * e.g:
+   *  encConfigPath = enc.conf
+   *  - Before : { enc.conf.secret_key : ENC(rOF43721f0pZqAXg#63a) }
+   *  - After  : { secret_key : decrypted_val }
+   *
+   * @param config
+   * @param encConfigPath
+   * @return
+   */
+  public static Config resolveEncrypted(Config config, Optional<String> encConfigPath) {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java]: What's the point of optional `encConfigPath` if we'll just return `config` if it is not present?

File: gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java [hunk 6]
```diff
@@ -462,4 +464,34 @@ public static boolean verifySubset(Config superConfig, Config subConfig) {
...
+    if (!encConfigPath.isPresent() || !config.hasPath(encConfigPath.get())) {
+      return config;
+    }
+
+    Config encryptedConfig = config.getConfig(encConfigPath.get());
+
+    PasswordManager passwordManager = PasswordManager.getInstance(configToProperties(config));
+    Map<String, String> tmpMap = Maps.newHashMap();
+    for (Map.Entry<String, ConfigValue> entry : encryptedConfig.entrySet()) {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java]: If the config is deep (e.g. `enc.conf.my.secret.key`) `encryptedConfig.entrySet()` will return a weird object for key `my`. What you want to do here instead is `ConfigUtils.configToProperties(encryptedConfig).entrySet()`.

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java [hunk 7]
```diff
@@ -110,6 +113,7 @@ public static Configuration newConfiguration() {
     // Add a new custom filesystem mapping
     conf.set("fs.sftp.impl", "gobblin.source.extractor.extract.sftp.SftpLightWeightFileSystem");
     conf.set("fs.sftp.impl.disable.cache", "true");
+    conf.set("fs.adl.impl", "org.apache.hadoop.fs.adl.AdlFileSystem");
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: This should not be needed. `FileSystem` uses service providers, so discovery happens automatically.

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java [hunk 8]
```diff
@@ -656,9 +660,37 @@ public static void safeRenameRecursively(FileSystem fileSystem, Path from, Path
...
+   *
+   * Common use case for these encryptedPath:
+   *   When there's have encrypted credential in job property but you'd like Filesystem to get decrypted value.
+   *
+   * @param srcConfig source config.
+   * @param encryptedPath Optional. If provided, config that is on this path will be decrypted. @see ConfigUtils.resolveEncrypted
+   *                      Note that config on encryptedPath will be included in the end result even it's not part of includeOnlyPath
+   * @return Hadoop Configuration.
+   */
+  public static Configuration getConfFromState(State state, Optional<String> encryptedPath) {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: Should we instead have a static path for encrypted keys used in Hadoop configurations?

File: gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemWrapper.java [hunk 9]
```diff
@@ -78,7 +78,7 @@ public void setProxiedFileSystem(FileSystem currentProxiedFs) {
    * @throws URISyntaxException
    * @return proxiedFs
    */
-  public FileSystem getProxiedFileSystem(State properties, AuthType authType, String authPath, String uri)
+  public FileSystem getProxiedFileSystem(State properties, AuthType authType, String authPath, String uri, final Configuration conf)
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemWrapper.java]: Please add a backwards compatible method (it can use this one internally).

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 2 above)
Comment (jinhyukchang) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Will do.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java (hunk 4 above)
Comment (jinhyukchang) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java]: Will remove it.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java (hunk 3 above)
Comment (jinhyukchang) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java]: Will do.

File: gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java (hunk 5 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java]: I think it's just two different pattern for caller:
 a) config = resolveEncrypted(config, Optional.fromNullable(val))
     or config = resolveEncrypted(config, Optional.absent())
 b) if (val != null) { config = resolveEncrypted(config); }

I chose a, but don't have strong preference. Thoughts?

File: gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java (hunk 6 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java]: When I tested out, encryptedConfig.entrySet() provided a full path on the key. Let me update the test case to test nested key as well.

File: gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemWrapper.java (hunk 9 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemWrapper.java]: Will do.

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java (hunk 8 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: You mean to have static path for Hadoop Config instead of getting it from caller?
The reason I've delegated this prefix out to the caller instead of having a static path in HadoopUtils is for the case of naming collision.
For example, if we need different credential for source Filesystem and target Filesystem, it would be hard to disambiguate using static path.
e.g:
File system accepts property "credential"
prefix for Source: src.encrypted.conf
prefix for Writer: writer.encrypted.conf

File: gobblin-core/build.gradle (hunk 1 above)
Comment (jinhyukchang) [inline gobblin-core/build.gradle]: This is to include two jars to use ADL FileSystem in runtime. Do we want to place it in different process?

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java (hunk 7 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: Yes, it supposed to pick it up in runtime without this configuration, but it's not working with AdlFileSystem. I looked at their jar and they seemed misplaced service information under "META-INF/org.apache.hadoop.fs.FileSystem", not in "META-INF/services/org.apache.hadoop.fs.FileSystem"

File: gobblin-core/build.gradle (hunk 1 above)
Comment (ibuenros) [inline gobblin-core/build.gradle]: Yes, this should not be in `gobblin-core`. You could create a module for ADL, but I expect it would be almost empty, otherwise, just make documentation specifying the artifacts that need to be added for running ADL pipeline.

File: gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java (hunk 5 above)
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java]: Sure, just make it clear that passing `Optional.absent()` will not change anything, as opposed to attempting to decrypt everything.

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java (hunk 7 above)
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: I see, if you add an ADL module you could instead use the service provider pattern with your own resource. Have you considered submitting a PR to their project fixing this?

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java (hunk 8 above)
Comment (ibuenros) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: Sounds good

File: gobblin-core/build.gradle (hunk 1 above)
Comment (jinhyukchang) [inline gobblin-core/build.gradle]: With the comment below regarding service provider, I think it's better to add a module for ADL. Will remove this line and add a module for ADL.

File: gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java (hunk 7 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/HadoopUtils.java]: Will add a ADL module as suggested. I am having trouble finding their github page. It's not part of hadoop project. Let me either do the PR or create an issue on this.

File: gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java (hunk 5 above)
Comment (jinhyukchang) [inline gobblin-utility/src/main/java/gobblin/util/ConfigUtils.java]: Sure, will mention it in Java doc.

**Comment by jinhyukchang** [comment]
@ibuenros Just addressed PR comments. Could you take a look when you have a chance? Thanks!

File: gobblin-distribution/build.gradle [hunk 10]
```diff
@@ -47,6 +47,7 @@ dependencies {
   compile project(':gobblin-yarn')
   compile project(':gobblin-aws')
   compile project(':gobblin-runtime-hadoop')
+  compile project(':gobblin-azure-datalake')
```
Comment (ibuenros) [inline gobblin-distribution/build.gradle]: This should not be part of gobblin-distribution. Instead, add it to `gobblin-distribution/gobblin-flavor-full.gradle`

Comment (jinhyukchang) [inline gobblin-distribution/build.gradle]: Will move it and will take a look at test failures.
